### ISSUE TITLE
Bug ngrx/avoid-cyclic-effects

### ISSUE THREAD

**Issue by srnec** [timdeschryver/eslint-plugin-ngrx#223]
## Behavior
I don't see any cyclic issue here but I am getting this rule violation from eslint.
I am not sure how internally it is evaluated such an effect like this. Maybe just a naming issue?

## Code snippet

```
    public postPortInData$ = createEffect(() =>
        this.actions$.pipe(
            ofType<PostPortInData>(OrderEntityActionTypes.postPortInData),
            map((action) => action.payload),
            concatLatestFrom(() => this.appStoreFacadeService.order.orderEntity$),
            switchMap(([orderContext, orderEntity]) =>
                this.orderEntityService.queueContextCall(orderContext, orderEntity).pipe(
                    map(() => new PostPortInDataSuccess()),
                    catchError(() => of(new ConfigErrorShowModal()))
                )
            )
        )
    );
```

## I would be willing to submit a PR to fix this issue :heart:

[ ] Yes (Assistance is provided if you need help submitting a pull request)
[x] No

**Comment by timdeschryver** [comment]
What version of our linter are you using?
Could you also share the actions definitions please, or a reproduction.

**Comment by timdeschryver** [comment]
I tried to reproduce this, and my guess is that one of the succes or error actions has a similar action type to  `postPortInData`

**Comment by timdeschryver** [comment]
Feel free to reopen with more information if that isn't the case @srnec

**Comment by srnec** [comment]
Hi Tim!

Sorry, I was too busy with pandemic related stuff.

here are the action definitions:
```
export enum OrderEntityActionTypes {
    postPortInData = '[Order Entity] Post PortIn Data',
    postPortInDataSuccess = '[Order Entity] Post PortIn Data Success',
    postPartnerData = '[Order Entity] Post provider Data',
}

export class PostPortInData implements Action {
    public readonly type: OrderEntityActionTypes = OrderEntityActionTypes.postPortInData;

    constructor(public payload: OrderContext) {}
}

export class PostPortInDataSuccess implements Action {
    public readonly type: OrderEntityActionTypes = OrderEntityActionTypes.postPortInDataSuccess;
}

export class PostPartnerData implements Action {
    public readonly type: OrderEntityActionTypes = OrderEntityActionTypes.postPartnerData;

    constructor(public payload: PartnerData) {}
}

export type OrderEntityActions = PostPortInData | PostPortInDataSuccess | PostPartnerData;
```

but I don't see anything wrong here, or something suspicious.

**Comment by timdeschryver** [comment]
I was also able to reproduce it now, seems like the culprit is the generic that is passed to the `ofType` operator.
I'll take a look if we can provide a fix for these cases.

**Comment by rafaelss95** [comment]
Should we reopen this then?

---

@srnec sorry, but out of curiosity, is there any reason to use the _old_ way of action creation?

@timdeschryver maybe it would be a good idea to create a rule for this..

**Comment by srnec** [comment]
> Should we reopen this then?

Yes we can
> @srnec sorry, but out of curiosity, is there any reason to use the _old_ way of action creation?

If the effort for the fixing this issue is too high then it is not worth but I can imagine there are huge projects that contain a lot of actions and reducers (yes, don't forget about reducers because based on my knowledge, if you switch to new actions you have to switch reducers where it is used as well). 

Short story short, I think it is worth to fix it if the effort for fixing is reasonable.

**Comment by timdeschryver** [comment]
I thought I had reopened this yesterday 😅
But yea, I agree... it could be harder than expected to get this resolved. If that's the case I think a note should be enough 😅

**Comment by rafaelss95** [comment]
I tried to reproduce this locally, but I failed. Do you mind to point me what I'm missing here? https://github.com/rafaelss95/eslint-plugin-ngrx/commit/298461db3b69c6f2176ece1d182c7f01e48fdd7a

Currently it report 0 errors (I put the error range randomly, just for test).

**Comment by timdeschryver** [comment]
@rafaelss95 this should do it (no `catchError` case, because an effect should always return an action so returning null makes it not compile).

```ts
import { Component, Injectable } from '@angular/core';
    import { Actions, createEffect, concatLatestFrom, ofType } from '@ngrx/effects';
    import { Action, createAction, Store } from '@ngrx/store';
    import { Observable, of } from 'rxjs';
    import { catchError, map, switchMap } from 'rxjs/operators';
    export enum OrderEntityActionTypes {
      postPortInData = '[Order Entity] Post PortIn Data',
      postPortInDataSuccess = '[Order Entity] Post PortIn Data Success',
      postPartnerData = '[Order Entity] Post provider Data',
    }
    type OrderContext = { id: string }
    type PartnerData = { name: string }
    export class PostPortInData implements Action {
      public readonly type: OrderEntityActionTypes =
        OrderEntityActionTypes.postPortInData
      constructor(public payload: OrderContext) {}
    }
    export class PostPortInDataSuccess implements Action {
      public readonly type: OrderEntityActionTypes =
        OrderEntityActionTypes.postPortInDataSuccess
    }
    export class PostPartnerData implements Action {
      public readonly type: OrderEntityActionTypes =
        OrderEntityActionTypes.postPartnerData
      constructor(public payload: PartnerData) {}
    }
    export type OrderEntityActions =
      | PostPortInData
      | PostPortInDataSuccess
      | PostPartnerData
    @Injectable()
    class OrderEntityService {
      queueContextCall(orderContext, orderEntity): Observable<unknown> {
        return of()
      }
    }
    @Injectable()
    class AppStoreFacadeService {
      order = { orderEntity$: of({}) }
    }
    @Injectable()
    class Effect {
      public postPortInData$ = createEffect(() =>
        this.actions$.pipe(
          ofType<PostPortInData>(OrderEntityActionTypes.postPortInData),
          map((action) => action.payload),
          concatLatestFrom(() => this.appStoreFacadeService.order.orderEntity$),
          switchMap(([orderContext, orderEntity]) =>
            this.orderEntityService
              .queueContextCall(orderContext, orderEntity)
              .pipe(
                map(() => new PostPortInDataSuccess()),
                catchError(() => of(null)),
              ),
          ),
        ),
      )
      constructor(
        private readonly actions$: Actions,
        private readonly appStoreFacadeService: AppStoreFacadeService,
        private readonly orderEntityService: OrderEntityService,
      ) {}
    }
```

**Comment by rafaelss95** [comment]
I hope the fix in #241 is correct 🙏 

About disallowing the `Action` classes, I wrote something in https://github.com/timdeschryver/eslint-plugin-ngrx/discussions/134.
